pep_sphinx_extensions/pep_processor/parsing/pep_role.py

```diff
@@ -1,14 +1,28 @@
+from docutils import nodes
 from sphinx import roles
 
 
-class PEPRole(roles.PEP):
+class PEPRole(roles.ReferenceRole):
     """Override the :pep: role"""
-    # TODO override the entire thing (internal should be True)
 
-    def build_uri(self) -> str:
-        """Get PEP URI from role text."""
+    def run(self) -> tuple[list[nodes.Node], list[nodes.system_message]]:
+        # Get PEP URI from role text.
         pep_str, _, fragment = self.target.partition("#")
-        pep_base = self.inliner.document.settings.pep_url.format(int(pep_str))
+        pep_num = int(pep_str)
+        pep_base = self.inliner.document.settings.pep_url.format(pep_num)
         if fragment:
-            return f"{pep_base}#{fragment}"
-        return pep_base
+            ref_uri = f"{pep_base}#{fragment}"
+        else:
+            ref_uri = pep_base
+        if self.has_explicit_title:
+            title = self.title
+        else:
+            title = f"PEP {pep_num}"
+        try:
+            return [nodes.reference("", title,
+                                    internal=True, refuri=ref_uri, classes=["pep"],
+                                    _title_tuple=(pep_num, fragment))], []
```

Comment:
Style nit—IMO, the current structure is potentially confusing, hard to read and follow, and I find it visually ugly. I suggest:

```suggestion
            reference_node = nodes.reference(
                    "", title, internal=True, refuri=ref_uri, classes=["pep"],
                    _title_tuple=(pep_num, fragment))
            return [reference_node], []
```

or even better,

```suggestion
            reference_node = nodes.reference(
                "",
                title,
                internal=True,
                refuri=ref_uri,
                classes=["pep"],
                _title_tuple=(pep_num, fragment),
            )
            return [reference_node], []
```